### ISSUE TITLE
Insane library upgrade ordering

### ISSUE THREAD

**Issue by l29ah** [gentoo-haskell/haskell-updater#11]
haskell-updater doesn't produce a proper order for the packages to be built in, and portage can't figure it out (as it thinks that all the libraries-to-be-rebuilt are present and in a good condition?), resorting to random-ish order and predictably failing at the first package due to the broken deps after ghc upgrade.

**Comment by solpeth** [comment]
Yes, it is a pain. But I only experience failures with `FEATURES=test` enabled. Is this your case? If not that's something new to me.

**Comment by l29ah** [comment]
No, tests aren't enabled.

**Comment by solpeth** [comment]
Okay. I'm out of my depth with `haskell-updater` so hopefully someone else can take a look.

**Comment by l29ah** [comment]
Seems like `haskell-updater -- --with-bdeps=y` does the trick.

**Comment by solpeth** [comment]
Good to know! Thanks for the tip!

**Comment by trofi** [comment]
That is usually a sign of circular dependencies (explicit or implicit). It should not happen for either USE=test or USE=-test. Both types of circular dependencies should be tracked down and cycles broken if possible. We don't have nice tools to do that though.

I usually brute-force my ways into the cycles like that:
1. pick first package that fails to build
2. try to rebuild it's depends, check what breaks to build
3. collect graph manually (usually it's not too many steps)

Once done spot the cycle.

**Comment by hololeap** [comment]
This is something the package manager needs to figure out. All `haskell-updater` does is throw a list of broken packages to the PM, and hopefully we have sensible options for its invocation.
